src/daemon/daemon.cpp

```diff
@@ -1625,136 +1695,47 @@ try // clang-format on
     mpl::ClientLogger<InfoReply, InfoRequest> logger{mpl::level_from(request->verbosity_level()), *config->logger,
                                                      server};
     InfoReply response;
+    InstanceSnapshotsMap instance_snapshots_map;
 
     // Need to 'touch' a report in the response so formatters know what to do with an otherwise empty response
     request->snapshot_overview() ? (void)response.mutable_snapshot_overview()
                                  : (void)response.mutable_detailed_report();
     bool have_mounts = false;
     bool deleted = false;
-    auto fetch_instance_info = [&](VirtualMachine& vm) {
+
+    auto fetch_detailed_report = [&](VirtualMachine& vm) {
+        fmt::memory_buffer errors;
         const auto& name = vm.vm_name;
-        auto info = response.mutable_detailed_report()->add_details();
-        auto instance_info = info->mutable_instance_info();
-        auto present_state = vm.current_state();
-        info->set_name(name);
-        if (deleted)
-        {
-            info->mutable_instance_status()->set_status(mp::InstanceStatus::DELETED);
-        }
-        else
-        {
-            info->mutable_instance_status()->set_status(grpc_instance_status_for(present_state));
-        }
 
-        auto vm_image = fetch_image_for(name, config->factory->fetch_type(), *config->vault);
-        auto original_release = vm_image.original_release;
+        const auto& it = instance_snapshots_map.find(name);
+        const auto& [pick, all] = it == instance_snapshots_map.end() ? SnapshotPick{{}, true} : it->second;
 
-        if (!vm_image.id.empty() && original_release.empty())
+        try
         {
-            try
-            {
-                auto vm_image_info = config->image_hosts.back()->info_for_full_hash(vm_image.id);
-                original_release = vm_image_info.release_title.toStdString();
-            }
-            catch (const std::exception& e)
+            if (all)
             {
-                mpl::log(mpl::Level::warning, category, fmt::format("Cannot fetch image information: {}", e.what()));
+                populate_instance_info(vm, response.mutable_detailed_report()->add_details(),
+                                       request->no_runtime_information(), deleted, have_mounts);
+                for (const auto& snapshot : pick)
+                    vm.get_snapshot(snapshot); // verify validity of any snapshot name requested separately
```

Comment:
I don't think you need this here, since you are iterating over the individual snapshot arguments in the next few lines.